### ISSUE TITLE
A function to rewire given edges manually while keeping the others

### ISSUE THREAD

**Issue by heavywatal** [igraph/igraph#2677]
**What is the feature or improvement you would like to see?**
Suppose you want to modify only `from` end of a given edge while keeping its `to` end (or want to modify both ends of the edge). To achieve this simple rewiring, you currently have to execute `igraph_delete_edges()` to remove the existing one and `igraph_add_edges()` to add the modified one. The internal graph data (`oi`, `ii`, `os`, `is`) are calculated twice. Many edges are assigned different IDs (indices). It would be very nice to have a function to modify edges more efficiently and gracefully. The interface will look like this:

```c++
igraph_error_t
igraph_modify_edges(igraph_t *graph, const igraph_es_t edges_to_mod, const igraph_vector_int_t *end_vertices);
```

**Use cases for the feature**
Replicating Watts–Strogatz model is an obvious example.
I am currently working on simulating recombination on gene genealogies from neutral evolution.

**References**
Watts, D., Strogatz, S. Collective dynamics of ‘small-world’ networks. Nature 393, 440–442 (1998). https://doi.org/10.1038/30918

**Comment by szhorvat** [comment]
As I understand, you are asking for such a function with a good complexity (quasilinear or similar) in the number of edges to modify.

I understand the need to do this (I've encountered it as well), but I am not sure that this is currently feasible. The problem is that the internal indices need to be re-sorted after every modification, so there is no way to avoid modifying the _entire_ graph data structure.

igraph's own rewiring functions work by converting the graph to a more application-suitable format, performing the rewiring, then converting back. This format is typically a non-sorted or partially sorted adjacency list.

I do plan to make significant improvements to the random rewiring functionality. What could in principle be done is to create a rewiring-specific graph representation and expose it.  The operations it should support efficiently include selecting an edge uniformly at random, modify the endpoints of an edge, and detect if an edge with given endpoints already exists.

While in an ideal situation the core data structure would support all these efficiently, we're facing tradeoffs here: an efficiency improvement in one area necessitates a concession in another. igraph has some features that other libraries (that support more efficient modification) struggle with, namely full multigraph support and the ability to treat edges as reversed (mode=IN, OUT). We did discuss redesigning the core data structure, but this is a large and time consuming project that we don't have the resources for at this moment. It will take years before we get there, assuming that igraph receives the required resources (funding, contributors), which is currently the biggest obstacle.

----

Are you intending to program this in R or in C?

I will ping you when I get to the rewiring work so we can talk about how we can make it usable for your purposes.

**Comment by heavywatal** [comment]
Thank you for your consideration and reply. I am writing an R package to use C igraph via cpp11. So the idea "to create a rewiring-specific graph representation and expose it" sounds promising. But it might be overkill to my current purpose, which is the preservation of edge list representation. The efficiency/performance was a hopeful byproduct for me. I think I still do not fully understand the data structure and granularity of "every modification". Here is my understanding (with a single edge rewiring for simplicity):

Current behavior with two steps:
1. `igraph_delete_edges(g, i)`:
   - core data (`from`, `to`) is modified, involving the change in the array size;
   - internal indices (`oi`, `ii`, `os`, `is`) are re-sorted;
   - as a result, the original i-th edge is removed from the edge list representation;
     the previous (i+1)-th edge becomes new i-th edge; the rest is shifted in the same way.
2. `igraph_add_edges(g, vertices)`:
   - core data is modified, involving the change in the array size;
   - internal indices are re-sorted;
   - as a result, the new edge of `vertices` is appended to the tail of the edge list representation.

My request, based on wild guess:
1. `igraph_modify_edges(g, i, vertices)`:
   - core data of i-th edge (`from[i]`, `to[i]`) is modified in-place (to `vertices[0]`, `vertices[1]`);
   - internal indices are re-sorted;
   - as a result, only the i-th edge (row) in the edge list representation is apparently changed.

If it is correct (I guess it is wrong because I miss something),
- the modification on the core data (and re-allocation of whole array as well?) may be reduced,
- the internal indices need to be re-sorted only once after all the modifications on the core data, and
- the edge list can be mostly preserved, no matter how drastically the internal indices are modified.
